Validate valueset schema

ryanbnl opened the pull request:
DRAFT - just making this for a discussion

ryanbnl commented:
@dslmeinte can you have a look - I left a TODO comment in the Makefile. Which approach do you think is best? I prefer "use default schema for all except x,y,z" and then provide a list of "x, y, z". But then I'd want to see if I can migrate the build to say package.json (need to investigate a little to see if we do it there in a nice and clean way).

dslmeinte commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+node_modules/
+package-lock.json
```
It's best to do commit this, as it pins the versions of dependencies (also transitively).

dslmeinte commented on test.js:
```diff
@@ -0,0 +1,43 @@
+];
+
+// Get a list of all of the valuesets
+valuesets = fs
```
*cough* `const` *cough* ;)

dslmeinte commented on test.js:
```diff
@@ -0,0 +1,43 @@
+// Take all files in [folderA] and run them against schema in [folderA] named [fileA].schema.json, or [default]
+
+const fs = require('fs');
+var exec = require('child_process').exec;
```
*cough* `const` *cough* ;)

dslmeinte commented on test.js:
```diff
@@ -0,0 +1,43 @@
+
+valuesets.forEach(item => {
+    // Decide which schema to use - the named schema or the default
+    let schema = schemas.find(x => x === item.replace(".json", ".schema.json")) || DEFAULT_SCHEMA;
```
*cough* `const` *cough* ;)

dslmeinte replied:
AJV can also be run as a JS-lib, rather than as a CLI-command. But we could keep it close to the approach in the `eu-dcc-schema` repo as well.

dslmeinte commented on test.js:
```diff
@@ -0,0 +1,43 @@
+        function (error, stdout, stderr) {
+            console.log(stdout);
+            if (error !== null) {
+                console.log('stderr: ' + stderr);
```
In case of failure, it'd make sense to `process.exit(1)` so that the GitHub Action fails - I don't think it does just from printing something to `stderr`. (Or does it?) (For the user it'd be most convenient if *all* value sets get reported on, not only the first one that fails.)

dslmeinte replied:
I checked: even with an error on `stderr`, the GitHub Action doesn't fail.

ryanbnl replied:
Thought so, I've always had to use an explicit exit code (works the same on Windows, it gets swallowed).

dslmeinte commented:
Looks good!

ryanbnl commented:
Moved over to using ajv directly, which also fixed an issue with the previous version at the same time (the shell was executed out of order causing the exit code to execute before the shells were finished).

dslmeinte commented on test.js:
```diff
@@ -0,0 +1,73 @@
+
+// Init ajv
+const ajv = new Ajv2020({
+    strict: false
```
Why non-strict? Options that I regularly use:
```javascript
const ajv = new Ajv({
    allErrors: true,        // don't stop after 1st error
    strict: true,
    allowUnionTypes: true,  // necessary for validating against the meta-schema
    validateSchema: false   // prevent that Ajv throws with 'no schema with key or ref "https://json-schema.org/draft/2020-12/schema"'
    // It seems weird that Ajv can't validate the schemas themselves (against a meta-schema) but you have to tweak things to do that yourself...
})
```

ryanbnl replied:
I'm using the same settings as in the original version.

dslmeinte replied:
That's a good thing. We could entertain adding `allErrors: true` to avoid that people (feel they have to) round-trip through a GitHub Action showing at most 1 error (per valueset).

dslmeinte commented:
It would be good to put a sentence or 2 in the README about the fact that the value sets are validated against schemas (and what the schemas are for, exactly, as we have 3).

After that: let's squash to a new branch, and merge a new PR.

dslmeinte commented on .github/workflows/test.yml:
```diff
@@ -13,5 +13,5 @@ jobs:
     - name: Install node dependencies
       run: npm install
-    - name: Initialise then run the tests
+    - name: Validate the valusets
```
(typo: "valu[e]sets")

ryanbnl commented:
@skounis do you mind if we rename this + eventually do a squash merge? the history is a bit... messy :)

ryanbnl commented:
(I'd like to rename to ("feat/validation-valuesets-schema")

skounis commented:
> @skounis do you mind if we rename this + eventually do a squash merge? the history is a bit... messy :)

That's fine, please go ahead.

dslmeinte commented:
The existence of the GitHub Action is not mentioned in the README.
